doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,968 @@
+
+Playtime is over
+****************
+
+The method we have to implement is ``runOnBasicBlock`` which takes as parameter a reference to the current block. We will proceed step by step.
+
+Finding null literals
++++++++++++++++++++++
+
+To find the null literals we need to iterate over every instruction of the block and check if one of the operands is null.
+
+.. code:: C++
+
+  //Add the following to your headers
+  #include "llvm/IR/Constants.h"
+  #include "llvm/IR/Instructions.h"
+
+  //Add the following to MyPass
+  virtual bool runOnBasicBlock(BasicBlock &BB) override {
+    // Not iterating from the beginning to avoid obfuscation of Phi instructions
+    // parameters
+    for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),
+                                       end = BB.end();
+         I != end; ++I) {
+      Instruction &Inst = *I;
+      // We are not using an iterator because we will need i later.
+      for (size_t i{0}; i < Inst.getNumOperands(); ++i) {
+        if (Constant *C = isValidCandidateOperand(Inst.getOperand(i))) {
+          errs() << "I've found one sir!\n";
+        }
+      }
+    }
+    return false;
+   }
+
+  Constant *isValidCandidateOperand(Value *V) {
+    if (Constant *C = dyn_cast<Constant>(V)) {
+      // Checking constant eligibility
+      if (isa<PointerType>(C->getType())) {
+        // dbgs() << "Ignoring NULL pointers\n";
+        return nullptr;
+      } else if (C->getType()->isFloatingPointTy()) {
+        // dbgs() << "Ignoring Floats 0\n";
+        return nullptr;
+      } else if (C->isNullValue()) {
+        return C;
+      } else {
+        return nullptr;
+      }
+    } else {
+      return nullptr;
+    }
+  }
+
+The ``runOnBasicBlock`` method is going to iterate through all the instructions of the block (for loop) and check if any operand of those instructions is an eligible null literal.
+If any of the operand is a null literal we print a message on the debug stream and we continue.
+You may have noticed the for loop is initialized with ``BB.getFirstInsertionPt()``.
+We could have iterated through the block with a foreach like:
+
+.. code:: C++
+
+    for(auto &I : BB) {
+    }
+
+But we do not want to modify the special instructions located at the begining of the block (the `phi instructions <http://en.wikipedia.org/wiki/Static_single_assignment_form#Converting_out_of_SSA_form>`_), so we skip them altogether and set the iterator to the first 'normal' instruction.
+
+The ``isValidCandidateOperand`` method checks if its parameter is a literal (constant means literal in LLVM, not variable declared ``const``).
+It also checks the type of the literal, it must not be a pointer or a floating point value (you will see later why).
+The type checks are done with the ``isa<>`` function which checks if its parameter can be cast to the type given by the template parameter.
```

Comment:
It's used in LLVM instead of RTTI(run time type information) because it was deemed to expensive http://llvm.org/docs/CodingStandards.html#do-not-use-rtti-or-exceptions